snow/engine/snowman/syncer/blocks_backfiller.go

```diff
@@ -0,0 +1,202 @@
+	outstandingRequests common.Requests     // tracks which validators were asked for which block in which requests
+	interrupted         bool                // flag to allow backfilling restart after recovering from validators disconnections
+}
+
+func NewBlockBackfiller(cfg BlockBackfillerConfig) *BlockBackfiller {
+	return &BlockBackfiller{
+		BlockBackfillerConfig: cfg,
+
+		fetchFrom:   set.Of[ids.NodeID](cfg.Validators.GetValidatorIDs(cfg.Ctx.SubnetID)...),
+		interrupted: len(cfg.Peers.PreferredPeers()) > 0,
+	}
+}
+
+func (bb *BlockBackfiller) Start(ctx context.Context) error {
+	ssVM, ok := bb.VM.(block.StateSyncableVM)
+	if !ok {
+		bb.Ctx.StateSyncing.Set(false)
+		return nil // nothing to do
+	}
+
+	switch wantedBlk, _, err := ssVM.BackfillBlocksEnabled(ctx); {
+	case err == block.ErrBlockBackfillingNotEnabled:
+		bb.Ctx.Log.Info("block backfilling not enabled")
+		bb.Ctx.StateSyncing.Set(false)
+		return nil
+	case err != nil:
+		return fmt.Errorf("failed checking if state sync block backfilling is enabled: %w", err)
+	default:
+		return bb.fetch(ctx, wantedBlk)
+	}
+}
+
+// Ancestors handles the receipt of multiple containers. Should be received in
+// response to a GetAncestors message to [nodeID] with request ID [requestID]
+func (bb *BlockBackfiller) Ancestors(ctx context.Context, nodeID ids.NodeID, requestID uint32, blks [][]byte) error {
+	// Make sure this is in response to a request we made
+	wantedBlkID, ok := bb.outstandingRequests.Remove(nodeID, requestID)
+	if !ok { // this message isn't in response to a request we made
+		bb.Ctx.Log.Debug("received unexpected Ancestors",
+			zap.Stringer("nodeID", nodeID),
+			zap.Uint32("requestID", requestID),
+		)
+		return nil
+	}
+
+	lenBlks := len(blks)
+	if lenBlks == 0 {
+		bb.Ctx.Log.Debug("received Ancestors with no block",
+			zap.Stringer("nodeID", nodeID),
+			zap.Uint32("requestID", requestID),
+		)
+
+		bb.markUnavailable(nodeID)
+
+		// Send another request for this
+		return bb.fetch(ctx, wantedBlkID)
+	}
+
+	// This node has responded - so add it back into the set
+	bb.fetchFrom.Add(nodeID)
+
+	if lenBlks > bb.AncestorsMaxContainersReceived {
+		blks = blks[:bb.AncestorsMaxContainersReceived]
+		bb.Ctx.Log.Debug("ignoring containers in Ancestors",
+			zap.Int("numContainers", lenBlks-bb.AncestorsMaxContainersReceived),
+			zap.Stringer("nodeID", nodeID),
+			zap.Uint32("requestID", requestID),
+		)
+	}
+
+	ssVM, ok := bb.VM.(block.StateSyncableVM)
+	if !ok {
+		return nil // nothing to do
+	}
+
+	switch nextWantedBlkID, _, err := ssVM.BackfillBlocks(ctx, blks); {
+	case err == block.ErrStopBlockBackfilling:
+		bb.Ctx.Log.Info("block backfilling done")
+		bb.Ctx.StateSyncing.Set(false)
+		return nil
+	case err != nil:
+		bb.Ctx.Log.Debug("failed to backfill blocks in Ancestors",
+			zap.Stringer("nodeID", nodeID),
+			zap.Uint32("requestID", requestID),
+			zap.Error(err),
+		)
+		return bb.fetch(ctx, wantedBlkID)
```

Comment:
we should probably limit the number of times we reissue a block request after vm failure and fail the engine as well.